executor/src/witgen/jit/compiler.rs

```diff
@@ -0,0 +1,539 @@
+    (row * {column_count} + column) as usize
+}}
+
+#[inline]
+fn index_known(global_offset: u64, local_offset: i32, column: u64) -> (u64, u64) {{
+    let column = column - {first_column_id};
+    let row = (global_offset as i64 + local_offset as i64) as u64;
+    let words_per_row = ({column_count} + 31) / 32;
+    (row * words_per_row + column / 32, column % 32)
+}}
+
+#[inline]
+fn get(data: &[FieldElement], global_offset: u64, local_offset: i32, column: u64) -> FieldElement {{
+    data[index(global_offset, local_offset, column)]
+}}
+
+#[inline]
+fn set(data: &mut [FieldElement], global_offset: u64, local_offset: i32, column: u64, value: FieldElement) {{
+    let i = index(global_offset, local_offset, column);
+    data[i] = value;
+}}
+
+#[inline]
+fn set_known(known: &mut [u32], global_offset: u64, local_offset: i32, column: u64) {{
+    let (known_idx, known_bit) = index_known(global_offset, local_offset, column);
+    known[known_idx as usize] |= 1 << (known_bit);
+}}
+
+#[repr(C)]
+enum LookupCell<'a, T> {{
+    /// Value is known (i.e. an input)
+    Input(&'a T),
+    /// Value is not known (i.e. an output)
+    Output(&'a mut T),
+}}
+
+#[repr(C)]
+struct WitgenFunctionParams {{
+    data: *mut FieldElement,
+    known: *mut u32,
+    len: u64,
+    row_offset: u64,
+    call_machine: fn(u64, *mut [LookupCell<'_, FieldElement>], usize) -> bool,
+}}
+    "#
+    ))
+}
+
+#[cfg(test)]
+mod tests {
+    use pretty_assertions::assert_eq;
+
+    use powdr_number::GoldilocksField;
+
+    use super::*;
+
+    #[test]
+    fn compile_util_code() {
+        if let Err(e) = compile_effects::<GoldilocksField>(0, 2, &[], &[]) {
+            println!("{e}");
+            panic!()
+        }
+    }
+
+    fn cell(column_name: &str, id: u64, row_offset: i32) -> Cell {
+        Cell {
+            column_name: column_name.to_string(),
+            row_offset,
+            id,
+        }
+    }
+
+    fn symbol(cell: &Cell) -> SymbolicExpression<GoldilocksField, Cell> {
+        SymbolicExpression::from_symbol(cell.clone(), None)
+    }
+
+    fn number(n: u64) -> SymbolicExpression<GoldilocksField, Cell> {
+        SymbolicExpression::from(GoldilocksField::from(n))
+    }
+
+    fn assignment(
+        cell: &Cell,
+        e: SymbolicExpression<GoldilocksField, Cell>,
+    ) -> Effect<GoldilocksField, Cell> {
+        Effect::Assignment(cell.clone(), e)
+    }
+
+    #[test]
+    fn simple_effects() {
+        let a0 = cell("a", 2, 0);
+        let x0 = cell("x", 0, 0);
+        let ym1 = cell("y", 1, -1);
+        let yp1 = cell("y", 1, 1);
+        let effects = vec![
+            assignment(&x0, number(7) * symbol(&a0)),
+            assignment(&ym1, symbol(&x0)),
+            assignment(&yp1, symbol(&ym1) + symbol(&x0)),
+            Effect::Assertion(Assertion {
+                lhs: symbol(&ym1),
+                rhs: symbol(&x0),
+                expected_equal: true,
+            }),
+        ];
```

Comment:
Should we have a slightly more complex example to be used in both tests? It would be nice to test each operation at least once as well as edge cases like field division and negative field elements.

Reply:
Not sure I really want to test every operator, but probably a good idea...